File: app/src/main/java/com/kelsos/mbrc/services/OnSubscribeReader.java
package com.kelsos.mbrc.services;

import android.text.TextUtils;
import java.io.BufferedReader;
import java.io.IOException;
import rx.Observer;
import rx.observables.SyncOnSubscribe;

public class OnSubscribeReader extends SyncOnSubscribe<BufferedReader, String> {
  private final BufferedReader reader;

  public OnSubscribeReader(BufferedReader reader) {
    this.reader = reader;
  }

  @Override protected BufferedReader generateState() {
    return this.reader;
  }

  @Override protected BufferedReader next(BufferedReader state, Observer<? super String> observer) {
    try {
      String line = reader.readLine();
      if (TextUtils.isEmpty(line)) {
        observer.onCompleted();
      } else {
        observer.onNext(line);
      }
    } catch (IOException e) {
      observer.onError(e);
    } return state;
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/services/OnSubscribeReader.kt
package com.kelsos.mbrc.services

import io.reactivex.ObservableEmitter
import io.reactivex.ObservableOnSubscribe
import timber.log.Timber
import java.io.BufferedReader
import java.net.Socket

class OnSubscribeReader(
    private val reader: BufferedReader,
    private val socket: Socket
) : ObservableOnSubscribe<ServiceCaller.ServiceMessage> {
  override fun subscribe(observer: ObservableEmitter<ServiceCaller.ServiceMessage>) = try {
    while (true) {
      val line = reader.readLine()
      if (line.isNullOrBlank()) {
        break
      }

      Timber.v("incoming -> %s", line)
      observer.onNext(ServiceCaller.ServiceMessage(line, socket))
    }
    observer.onComplete()
  } catch (e: Throwable) {
    observer.onError(e)
  }
}
